File: Quick_Sort/quicksort.c
#include <stdio.h>
void swap(int* m, int* n)
{
    int x =*m;
    *m =*n;
    *n =x;
}

int partition (int arr[], int low, int high)
{
    int pivot = arr[high];   
    int i =low-1; 
 
    for (int j = low; j <= high- 1; j++)
    {
        if (arr[j] <= pivot)
        {
            i++;  
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}
void quick_Sort(int arr[], int low, int high)
{
    if (low < high)
    {
        int pi = partition(arr, low, high);
        quick_Sort(arr, low, pi - 1);
        quick_Sort(arr, pi + 1, high);
    }
}
void printarr(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("n");
}
 
int main()
{
    int arr[] = {10,9,8,7,6,5,4,3,2,1};
    int n = 10;
    quick_Sort(arr, 0, n-1);
    printf("Sorted array: n");
    printarr(arr, n);
    return 0;
}

File: algorithms/divide_conquer/quick_sort/quick_sort.rs
//  divide conquer | quick sort | Rust
//  Part of Cosmos by OpenGenus Foundation

fn quick_sort(mut arr :Vec<i32>,low :usize, high :usize) -> Vec<i32> {

    if low < high {
        let mid = partition(&mut arr, low, high);

        let arr = quick_sort(arr.clone(),low,mid);
        let arr = quick_sort(arr,mid+1,high);
        return arr
    }
    arr
}

fn partition(arr : &mut Vec<i32>, low :usize, high :usize) -> usize {
    let pivot = arr[low];
    let mut i = low ;
    let mut j = high;

    loop {
        while arr[i] < pivot && arr[i] != pivot {
            i += 1;
        }
        while arr[j] > pivot && arr[j] != pivot {
            j -= 1;
        }
        if i < j {
           arr.swap(i,j);
        }else{
            return j
        }
    }
}

fn main() {
    let arr = vec![10, 7, 8, 9, 1, 5];
    let len = arr.len();
    let sorted = quick_sort(arr,0,len-1);
    println!("Sorted array is {:?}", sorted);
}
